starknet/transaction.go

```diff
@@ -104,22 +104,59 @@ func (t *TransactionType) UnmarshalJSON(data []byte) error {
+
+func (r *Resource) UnmarshalJSON(data []byte) error {
+	switch string(data) {
+	case `"L1_GAS"`:
+		*r = ResourceL1Gas
+	case `"L2_GAS"`:
+		*r = ResourceL2Gas
+	default:
+		return fmt.Errorf("unknown resource: %s", string(data))
+	}
+	return nil
+}
+
+type DataAvailabilityMode uint32
+
+const (
+	DAModeL1 DataAvailabilityMode = iota
+	DAModeL2
+)
+
+type ResourceBounds struct {
+	MaxAmount       uint64     `json:"max_amount"`
+	MaxPricePerUnit *felt.Felt `json:"max_price_per_unit"`
+}
+
 // Transaction object returned by the feeder in JSON format for multiple endpoints
 type Transaction struct {
-	Hash                *felt.Felt      `json:"transaction_hash,omitempty" copier:"must,nopanic"`
-	Version             *felt.Felt      `json:"version,omitempty"`
-	ContractAddress     *felt.Felt      `json:"contract_address,omitempty"`
-	ContractAddressSalt *felt.Felt      `json:"contract_address_salt,omitempty"`
-	ClassHash           *felt.Felt      `json:"class_hash,omitempty"`
-	ConstructorCallData *[]*felt.Felt   `json:"constructor_calldata,omitempty"`
-	Type                TransactionType `json:"type,omitempty"`
-	SenderAddress       *felt.Felt      `json:"sender_address,omitempty"`
-	MaxFee              *felt.Felt      `json:"max_fee,omitempty"`
-	Signature           *[]*felt.Felt   `json:"signature,omitempty"`
-	CallData            *[]*felt.Felt   `json:"calldata,omitempty"`
-	EntryPointSelector  *felt.Felt      `json:"entry_point_selector,omitempty"`
-	Nonce               *felt.Felt      `json:"nonce,omitempty"`
-	CompiledClassHash   *felt.Felt      `json:"compiled_class_hash,omitempty"`
+	Hash                  *felt.Felt                   `json:"transaction_hash,omitempty" copier:"must,nopanic"`
+	Version               *felt.Felt                   `json:"version,omitempty"`
+	ContractAddress       *felt.Felt                   `json:"contract_address,omitempty"`
+	ContractAddressSalt   *felt.Felt                   `json:"contract_address_salt,omitempty"`
+	ClassHash             *felt.Felt                   `json:"class_hash,omitempty"`
+	ConstructorCallData   *[]*felt.Felt                `json:"constructor_calldata,omitempty"`
+	Type                  TransactionType              `json:"type,omitempty"`
+	SenderAddress         *felt.Felt                   `json:"sender_address,omitempty"`
+	MaxFee                *felt.Felt                   `json:"max_fee,omitempty"`
+	Signature             *[]*felt.Felt                `json:"signature,omitempty"`
+	CallData              *[]*felt.Felt                `json:"calldata,omitempty"`
+	EntryPointSelector    *felt.Felt                   `json:"entry_point_selector,omitempty"`
+	Nonce                 *felt.Felt                   `json:"nonce,omitempty"`
+	CompiledClassHash     *felt.Felt                   `json:"compiled_class_hash,omitempty"`
+	ResourceBounds        map[Resource]*ResourceBounds `json:"resource_bounds,omitempty"`
+	Tip                   uint64                       `json:"tip,omitempty"`
```

Comment:
I think 0 tip is still a valid value for v3 txns. So I believe it should be an `*uint64`

Reply:
[It is](https://external.integration.starknet.io/feeder_gateway/get_block?blockNumber=319100). But not sure why we need `uint64` to be a pointer in that case. Feeder always includes the tip, even if it's `0x0`.

Reply:
Tip is marked `omitempty` so if we ever marshal this type to JSON, it will get omitted if the value is 0.

Reply:
Just realized that I inadvertently fixed this when I changed `Tip` to a `*felt.Felt`, since the feeder formats it as a `felt`.